lua/diffview/vcs/adapters/hg/rev.lua

```diff
@@ -0,0 +1,52 @@
+local oop = require("diffview.oop")
+local Rev = require('diffview.vcs.rev').Rev
+local RevType = require('diffview.vcs.rev').RevType
+
+local M = {}
+
+---@class HgRev : Rev
+local HgRev = oop.create_class("HgRev", Rev)
+
+HgRev.NULL_TREE_SHA = "0000000000000000000000000000000000000000"
+
+function HgRev:init(rev_type, revision, track_head)
+  local t = type(revision)
+
+  assert(
+    revision == nil or t == "string" or t == "number",
+    "'revision' must be one of: nil, string, number!"
+  )
+  if t == "string" then
+    assert(revision ~= "", "'revision' cannot be an empty string!")
+  end
+
+  t = type(track_head)
+  assert(t == "boolean" or t == "nil", "'track_head' must be of type boolean!")
+
+  self.type = rev_type
+  self.track_head = track_head or false
+
+  self.commit = revision
+end
+
+function HgRev:object_name(abbrev_len)
+  if self.commit then
+    if abbrev_len then
+      return self.commit:sub(1, abbrev_len)
+    end
+
+    return self.commit
+  end
+
+  return "UNKNOWN"
+end
+
+function HgRev.to_range(rev_from, rev_to)
+  if rev_to then
+    return rev_from .. "::" .. rev_to
+  end
+  return rev_from
```

Comment:
In hg we use a revset which can be one or more commits:

-  `hg log -r <rev>`: will give the log of that commit only
- `hg log -r ::<rev>`: gives all commits from start (commit 0) to rev
- `hg log -r <rev>::`: gives all commits from rev until last commit
Also, there is a difference between (`<rev a>::<rev b>` and `<rev a>:<rev b>`, the latter will give all commits even on different branches, the first only on the current branch, where branch is a sequence of commits, not necessarily a named branch)

If you do `hg diff -r <rev>`, you will get the diff between `<rev>` and the current revision. `hg diff -r <rev a>::<rev b>` will give the diff between the two revisions.

Since there is in most cases no implicit range, I thought it would be better to leave it to the user?

Reply:
> - `hg log -r <rev>`: will give the log of that commit only
> - `hg diff -r <rev>`, you will get the diff between `<rev>` and the current revision

Right, it's this type of disparity that `to_range()` is supposed to circumvent. Would it work to do something like this maybe:

```
<rev>^::<rev>
```

Reply:
Oh, I think I understand now: you always want a range, even though it is a single change.

Reply:
In hg I use `hg diff -c <rev>` to the the change this commit introduces but that is just a shorthand for `<rev>^::<rev>` indeed

Reply:
No, I was wrong.
`hg diff -r` is not documented. It works as described above, but `hg log` does not.

Seems like I need to use `--from <rev> --to <rev>` for `diff`.

I'll need to do some more tests to make sure this is always giving me the right data.

Reply:
It took me a bit longer than I'd like to admit, but I think I figured it out.
@sindrets: just to confirm this is the right behaviour:
  * `DiffviewOpen` shows the uncommitted changes (there is no staging in hg)
  * `DiffviewOpen 3` shows the diff between revision 3 up to and including uncommitted changes
  * `DiffviewOpen 3::5` shows the diff between revision 3 up to and including  5 (i.e., changes introduced in 4 and 5)

Reply:
Yes, that's correct.